concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/GradientLimit.java

```diff
@@ -180,112 +212,84 @@ public static GradientLimit newDefault() {
         
-        if (sample.getCandidateRttNanos() < minRttThreshold) {
+        final long rttSample = sample.getCandidateRttNanos();
+        minWindowRttSampleListener.addSample(rttSample);
+        if (rttSample < minRttThreshold) {
             return;
         }
-        
-        final long rtt = sample.getCandidateRttNanos(); // rttWindowNoLoad.get().longValue();
-        minWindowRttSampleListener.addSample(rtt);
 
         final double queueSize = this.queueSize.apply((int)this.estimatedLimit);
         queueSizeSampleListener.addSample(queueSize);
 
-        // Reset or probe for a new noload RTT and a new estimatedLimit.  It's necessary to cut the limit
-        // in half to avoid having the limit drift upwards when the RTT is probed during heavy load.
-        // To avoid decreasing the limit too much we don't allow it to go lower than the queueSize.
-        if (probeMultiplier != DISABLED && resetRttCounter-- <= 0) {
-            resetRttCounter = nextProbeCountdown();
-            
-            estimatedLimit = Math.max(minLimit, Math.max(estimatedLimit - queueSize, queueSize));
-            rttNoLoad.update(current -> rtt);
-            LOG.debug("Probe MinRTT limit={}", getLimit());
-            return;
-        } else if (rttNoLoad.add(rtt)) {
-            LOG.debug("New MinRTT {} limit={}", TimeUnit.NANOSECONDS.toMicros(rtt)/1000.0, getLimit());
-        }
+        final double rttNoLoad = rttNoLoadAccumulator.add(rttSample).doubleValue();
+        final double rtt = (double)rttSample;
         
-        minRttSampleListener.addSample(rttNoLoad.get());
+        minRttSampleListener.addSample(rttNoLoad);
         
         final double gradient;
-        // rtt is still higher than rtt_noload because of smoothing rtt noload updates
+        // rtt is lower than rtt_noload because of smoothing rtt noload updates
         // set to 1.0 to indicate no queueing
-        if (rttNoLoad.get().doubleValue() > rtt) {
+        if (rtt < rttNoLoad) {
             gradient = 1.0;
         } else {
-            gradient = Math.max(0.5, rttTolerance * rttNoLoad.get().doubleValue() / rtt);
+            gradient = Math.max(0.5, rttNoLoad / rtt);
         }
         
         double newLimit;
         // Reduce the limit aggressively if there was a drop
         if (sample.didDrop()) {
             newLimit = estimatedLimit/2;
-        // Don't grow the limit because we are app limited
-        } else if ((estimatedLimit - sample.getMaxInFlight()) > queueSize) {
-            return;
         // Normal update to the limit
         } else {
             newLimit = estimatedLimit * gradient + queueSize;
         }
         
-        newLimit = Math.max(queueSize, Math.min(maxLimit, newLimit));
-        newLimit = Math.max(minLimit, estimatedLimit * (1-smoothing) + smoothing*(newLimit));
-        
         if ((int)newLimit != (int)estimatedLimit) {
```

Comment:
Just wondering whether this int casting may lead to undesirable false outcome (i.e. 20.999.. and 21.000..1 ), perhaps best to compare double with a small epsilon.